cmd/gin-repod/repo.go

```diff
@@ -630,3 +643,116 @@ func (s *Server) browseRepo(w http.ResponseWriter, r *http.Request) {
 
 	s.objectToWire(w, repo, obj)
 }
+
+// patchRepoSettings patches repository description and public status.
+// The request header has to contain an authorization header with a valid token.
+// The request body has to contain valid JSON containing "description" as key
+// and a string as value, "public" as key and a boolean value as value or both.
+// A request that does not contain any of these two keys will result in a BadRequest status.
+func (s *Server) patchRepoSettings(w http.ResponseWriter, r *http.Request) {
+	header := r.Header.Get("Authorization")
+	if header == "" || !strings.HasPrefix(header, "Bearer ") {
+		w.WriteHeader(http.StatusForbidden)
+		return
+	}
+
+	ivars := mux.Vars(r)
+	rid, err := s.varsToRepoID(ivars)
+
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	// Return StatusBadRequest if an error occurs or if the repository does not exist.
+	// Returning StatusNotFound for non existing repositories could lead to inference
+	// of private repositories later on.
+	exists, err := s.repos.RepoExists(rid)
+	if err != nil || !exists {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	_, ok := s.checkAccess(w, r, rid, store.AdminAccess)
+	if !ok {
+		return
+	}
+
+	// Pointer fields are used to make sure, that boolean values are
+	// not updated, when the corresponding field was not
+	// included in the submitted JSON.
+	var patch struct {
+		Description *string `json:"description,omitempty"`
+		Public      *bool   `json:"public,omitempty"`
+	}
+
+	if r.Body == nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	b, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	err = json.Unmarshal(b, &patch)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	var respBody string
+	if patch.Description == nil && patch.Public == nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	} else if patch.Description != nil && patch.Public != nil {
```

Comment:
The flow is a bit confusing here. Wouldn't it be more elegant to always return the whole repo data on success and then check for the presence of any patched fields and patch them, and then return all the repo data?